File: Data Structures & Algorithms/climbing-stairs/submission-3.java
class Solution {

    int cache[];

    public int climbStairs(int n) {
        cache = new int[n];
        for ( int i = 0; i < n; i++ ) {
            cache[i] = -1;
        }
        return dfs(n,0);
    }

    private int dfs(int n, int i) {

        // Base case
        if ( i >= n ) return ( i == n ) ? 1 : 0;

        // Recursive case
        if (cache[i] != -1) return cache[i];
        cache[i] = dfs(n, i + 1) + dfs(n, i + 2);
        return cache[i];

    }
}

File: Data Structures & Algorithms/valid-tree/submission-3.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        
        if len(edges) != n - 1:
            return False

        preMap = {i : [] for i in range(n)}
        for u, v in edges:
            preMap[u].append(v)
            preMap[v].append(u)

        visited = set()

        def dfs(node, prev):

            # Base Case
            if node in visited:
                return False

            # Recursive Case
            visited.add(node)
            for nei in preMap[node]:
                if nei == prev:
                    continue
                if not dfs(nei, node):
                    return False
            return True

        return dfs(0,-1) and len(visited) == n